R/germline.R

```diff
@@ -0,0 +1,109 @@
+#' This function creates germlines for clonal lineages. B cell clonal lineage represents a set of B cells
+#' that presumably have a common origin (arising from the same VDJ rearrangement event) and a common ancestor.
+#' Each clonal lineage has its own germline sequence that represents the ancestral sequence
+#' for each BCR in clonal lineage. In other words, germline sequence is a sequence of B-cells immediately
+#' after VDJ recombination, before B-cell maturation and hypermutation process. Germline sequence is useful
+#' for assessing the degree of mutation and maturity of the repertoire.
+#'
+#' @concept germline
+#'
+#' @aliases repGermline germline_single_df take_first_allele extract_gene_name generate_germline_sequence load_reference_sequences
+#'
+#' @importFrom stringr str_sub
+#' @importFrom purrr modify
+#' @importFrom magrittr %>% %<>%
+#' @importFrom VDJgermlines extractSequencesR
+
+#' @description Creates germlines for clonal lineages
+#'
+#' @usage
+#'
+#' repGermline(.data)
+#'
+#' @param .data The data to be processed. Can be \link{data.frame}, \link{data.table}
+#' or a list of these objects.
+#'
+#' It must have columns in the immunarch compatible format \link{immunarch_data_format}
+#'
+#' @return
+#'
+#' Data with added columns V.allele and J.allele (with first alleles of the genes),
+#' and V.sequence and J.sequence (with germline sequences).
+#'
+#' @examples
+#'
+#' data(immdata)
+#' repGermline(immdata$data)
+#' @export repGermline
+repGermline <- function(.data) {
+  if (inherits(.data, "list")) {
+    .validate_repertoires_data(.data)
+    .data %>%
+      lapply(function(sample_data) {
+        sample_data %>%
+          as_tibble() %>%
+          germline_single_df() %>%
+          return()
+      }) %>%
+      return()
+  } else {
+    .data %>%
+      as_tibble() %>%
+      germline_single_df() %>%
+      return()
```

Comment:
Did you check this with all data formats immunarch works with? Including db connections

Reply:
Do we have guides how to test it locally with different formats? Here I must convert data to dataframe to merge with reference dataframe; I think if as_tibble() doesn't work with some of the supported formats, we need to make common conversion function in tools.R.

Reply:
Found data formats in our tutorial: https://immunarch.com/articles/v2_data.html
It mentions MonetDBLite, which is deprecated and removed from CRAN; modern analog is duckDB. Both duckDB and Apache Spark can provide access with `tbl` function from `dplyr`; and this table can be converted to dataframe with `as_tibble()`, so this function will work with `dplyr::tbl()` compatible databases.